Exercises/chapter7/e1-a-robot.js

```diff
@@ -0,0 +1,192 @@
+  };
+}
+
+function runRobot(state, robot, memory) {
+  for (let turn = 0;; turn++) {
+    if (state.parcels.length === 0) {
+      return turn;
+    }
+    let action = robot(state, memory);
+    state = state.move(action.direction);
+    memory = action.memory;
+
+    if (turn > 50) {
+      console.log(state);
+      return;
+    }
+    // console.log(state);
+    //console.log(`Moved to ${action.direction}`);
+  }
+}
+
+function randomPick(array) {
+  let choice = Math.floor(Math.random() * array.length);
+  return array[choice];
+}
+function findRoute(graph, from, to) {
+  let work = [{at: from, route: []}];
+  for (let i = 0; i < work.length; i++) {
+    let {at, route} = work[i];
+    for (let place of graph[at]) {
+      if (place === to) return route.concat(place);
+      if (!work.some(w => w.at === place)) {
+        work.push({at: place, route: route.concat(place)});
+      }
+    }
+  }
+}
+
+function randomRobot(state) {
+  return {direction: randomPick(roadGraph[state.place])};
+}
+function routeRobot(state, memory) {
+  const mailRoute = [
+    "Alice's House", "Cabin", "Alice's House", "Bob's House",
+    "Town Hall", "Daria's House", "Ernie's House",
+    "Grete's House", "Shop", "Grete's House", "Farm",
+    "Marketplace", "Post Office"
+  ];
+
+  if (memory.length === 0) {
+    memory = mailRoute;
+  }
+  return {direction: memory[0], memory: memory.slice(1)};
+}
+function goalOrientedRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    let parcel = parcels[0];
+    if (parcel.place != place) {
+      route = findRoute(roadGraph, place, parcel.place);
+    } else {
+      route = findRoute(roadGraph, place, parcel.address);
+    }
+  }
+  return {direction: route[0], memory: route.slice(1)};
+}
+
+/**
+  This robot moves towards the closest parcel until all parcels are collected.
+  It then visits the closest address of each undelivered parcel until all are
+  delivered.
+*/
+function closestParcelsRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    if (parcels.some(p => p.place != place)) {
+      let closestParcel = parcels
+        .filter(p => p.place !== place)
+        .map(parcel => {
+          route = findRoute(roadGraph, place, parcel.place);
+          return route
+        })
+        .reduce((r1, r2) => r1.length <= r2 ? r1 : r2);
+      route = closestParcel;
+    } else {
+      let closestAddress = parcels.map(parcel => {
+        route = findRoute(roadGraph, place, parcel.address);
+        return route
+      }).reduce((r1, r2) => r1.length <= r2 ? r1 : r2);
+      route = closestAddress;
+    }
+  }
+  return {direction: route[0], memory: route.slice(1)};
+}
+
+/**
+  This robot moves towards the closer of the closest parcel and closest address
+  of collected parcels until all parcels are delivered.
+*/
+function closestParcelOrAddressRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    route = parcels
+      .map(parcel => {
```

Comment:
if you pull this lambda out into a function with a clear name it will be a lot easier to understand this code